### PR TITLE
Resolve animation issues with Date picker

### PR DESCRIPTION
**Pull request by disc0infern0**
* Refactored date/time pickers to new view and updated control of state with new enum, allowing animations of picker view transitions to be handled precisely with new function setPickerState.
* No settings for transitions or animations work well (in a Form) when transitioning from a state where the date picker is shown, so use the setPickerState function to make transitions from a date picker shown state without animation, but animate all other view transitions.
* There is now no janky animation whatsoever, but 2 out of the 6 possible transitions now happen without animation.

### CODE REVIEW

File: code/frontend/MakeItSo/MakeItSo.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -7,6 +7,8 @@
 	objects = {
 
 /* Begin PBXBuildFile section */
+		49281E992760EBCB0046A465 /* ReminderDateTimeView.swift in Sources */ = {isa = PBXBuildFile; fileRef = 49281E982760EBCB0046A465 /* ReminderDateTimeView.swift */; };
+		49281E9A2760EBCB0046A465 /* ReminderDateTimeView.swift in Sources */ = {isa = PBXBuildFile; fileRef = 49281E982760EBCB0046A465 /* ReminderDateTimeView.swift */; };
```
Comment (peterfriese) [inline code/frontend/MakeItSo/MakeItSo.xcodeproj/project.pbxproj:11]: Duplicate entries, might be a result of manually merging this file.

File: code/frontend/MakeItSo/Shared/Features/ReminderList/ViewModels/ReminderDetailsViewModel.swift [hunk 2]
```diff
@@ -129,12 +134,11 @@ class ReminderDetailsViewModel: ObservableObject {
     set {
       if newValue == true {
         reminder.dueDate = Date()
-        isShowingDatePicker = true
-      }
-      else {
-        hasDueTime = false
+        setPickerState(.date)
+      } else {
```
Comment (peterfriese) [inline code/frontend/MakeItSo/Shared/Features/ReminderList/ViewModels/ReminderDetailsViewModel.swift:138]: Personally, I am a fan of "egyptian style":

if condition {
  // do stuff
}
else { // <-- else is on new line, and (most importantly) parallel with the initial `if`
  // do other stuff
}

File: code/frontend/MakeItSo/Shared/Features/ReminderList/ViewModels/ReminderDetailsViewModel.swift [hunk 3]
```diff
@@ -148,27 +152,45 @@ class ReminderDetailsViewModel: ObservableObject {
...
+  // so define an enum to control view state of the pickers
+  @Published var pickerState: PickerState = .none
   
-  func toggleTimePicker() {
-    isShowingTimePicker.toggle()
-    isShowingDatePicker = false
+  func setPickerState(_ newvalue: PickerState) {
+    // Dont animate if transitioning from state where date picker is shown
+    if pickerState == .date {
+      pickerState = newvalue
```
Comment (peterfriese) [inline code/frontend/MakeItSo/Shared/Features/ReminderList/ViewModels/ReminderDetailsViewModel.swift:171]: Use camel case: newValue

File: code/frontend/MakeItSo/Shared/Features/ReminderList/Views/ReminderDateTimeView.swift [hunk 4]
```diff
@@ -0,0 +1,103 @@
+//
+//	ReminderDateTimeView.swift
```
Comment (peterfriese) [inline code/frontend/MakeItSo/Shared/Features/ReminderList/Views/ReminderDateTimeView.swift:2]: Does this tab come from my template file...?

File: code/frontend/MakeItSo/Shared/Features/ReminderList/Views/ReminderDateTimeView.swift [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+      timeToggleLabel
+    }
+    if viewModel.pickerState == .time {
+      DatePicker("Time", selection: $viewModel.dueDate, displayedComponents: .hourAndMinute)
+        .datePickerStyle(.wheel)
+    }
+  }
+
+  // Toggle label subviews for body
+  var dateToggleLabel: some View {
```
Comment (peterfriese) [inline code/frontend/MakeItSo/Shared/Features/ReminderList/Views/ReminderDateTimeView.swift:46]: Can probably refactor this even more by turning into a function and passing in the icon and title.

Rename the function to something like `expandableRow` or `expandableRowLabel`, and replace `timeToggleLable` entirely.

**Review by peterfriese — CHANGES_REQUESTED** [review]
A few comments regarding formatting, code structure, etc. Will take a closer look at the functionality later today.

File: code/frontend/MakeItSo/Shared/Features/ReminderList/Views/ReminderDateTimeView.swift (hunk 4 above)
Comment (disc0infern0) [reply code/frontend/MakeItSo/Shared/Features/ReminderList/Views/ReminderDateTimeView.swift:2]: Yes, I didn't change anything regarding templates, and that's what is generated from adding a new file.

File: code/frontend/MakeItSo/Shared/Features/ReminderList/Views/ReminderDateTimeView.swift (hunk 5 above)
Comment (disc0infern0) [reply code/frontend/MakeItSo/Shared/Features/ReminderList/Views/ReminderDateTimeView.swift:46]: I thought about it, but you would to pass five parameters actually:
icon, title, dateValue, handler, textStyle

 HStack {
      Image(icon)
        .frame(width: 26, height: 26) //, alignment: .center)
        .background(.blue)
        .foregroundColor(.white)
        .cornerRadius(4)
      VStack(alignment: .leading) {
        Text(title)
        if dateValue != nil {
          Button(action: handler) {
            Text(dateValue, style: style)
        } ...

Not worth it for something that likely won't be re-used again?

File: code/frontend/MakeItSo/MakeItSo.xcodeproj/project.pbxproj (hunk 1 above)
Comment (disc0infern0) [reply code/frontend/MakeItSo/MakeItSo.xcodeproj/project.pbxproj:11]: agreed.

File: code/frontend/MakeItSo/Shared/Features/ReminderList/ViewModels/ReminderDetailsViewModel.swift (hunk 2 above)
Comment (disc0infern0) [reply code/frontend/MakeItSo/Shared/Features/ReminderList/ViewModels/ReminderDetailsViewModel.swift:138]: Stylistically, I like that too, but generally I look to try to shorten my code as long as it remains clearly understood.
I am a fan of the old school Nassi Schneiderman style of coding, which I have adapted to mean that generally a function/view/discrete piece of code should all fit on the screen. If it doesn't, it needs refactoring.

Comment (disc0infern0) [reply code/frontend/MakeItSo/Shared/Features/ReminderList/ViewModels/ReminderDetailsViewModel.swift:138]: Appreciate your stylist conventions should take precedence in your repo!

File: code/frontend/MakeItSo/Shared/Features/ReminderList/ViewModels/ReminderDetailsViewModel.swift (hunk 3 above)
Comment (disc0infern0) [reply code/frontend/MakeItSo/Shared/Features/ReminderList/ViewModels/ReminderDetailsViewModel.swift:171]: agreed

File: code/frontend/MakeItSo/MakeItSo.xcodeproj/project.pbxproj (hunk 1 above)
Comment (disc0infern0) [reply code/frontend/MakeItSo/MakeItSo.xcodeproj/project.pbxproj:11]: On closer inspection, the two lines are not duplicates:  line 10 starts: 49281E99  vs line 11 starting: 49281E9A
The last character is different. Just to be sure, I ran 'uniq -d project.pbxproj' from the command line, and there was no output - indicating no duplicates.

**Review by peterfriese — CHANGES_REQUESTED** [review]
Just had an opportunity to check out how this looks like on the Simulator:

https://user-images.githubusercontent.com/232107/145356980-e555fd2b-ab81-4825-ae90-6b49874d7782.mp4

As becomes obvious from the screen recording, this is nowhere near the smooth transitions of the original app.

As long as we're not able to get close to the original, I am very reluctant to make wide-ranging changes to the code that will make it more difficult to read. Plus, there _is_ hope that Apple might actually fix this in the SDK itself, and I'd like to prevent us having to roll back a lot of changes we only made to workaround an issue.

Feel free to keep this PR open and continue optimising it, and use it as a reference when filing feedback to the SwiftUI team.

File: code/frontend/MakeItSo/MakeItSo.xcodeproj/project.pbxproj (hunk 1 above)
Comment (Hminchae) [reply code/frontend/MakeItSo/MakeItSo.xcodeproj/project.pbxproj:11]: agreed

**Pull request closed without merging**